### ISSUE TITLE
[Correctness][Ascend910B3][CANN 8.5][Triton 3.2] same-program output reload returns corrupted values

### ISSUE THREAD

**Issue by luentong** [triton-lang/triton-ascend#1416]
## Summary

We found an Ascend correctness failure in a FlagGems Triton top-k kernel.

With `renormalize=True`, a single same-program reload of a value that was just
written to `output_ptr` returns corrupted and nondeterministic data. Top-k ids
remain correct. Moving normalization into a second Triton kernel fixes the
identical workload.

The failure reproduces through both:

- the vLLM-Plugin-FL `fused_topk` wrapper;
- a direct call to `flag_gems.fused.topk_softmax`.

This report includes a minimal reproducer and compiler artifacts. We would
appreciate help determining whether the problem is in the Ascend backend,
`triton-adapter-opt`, `bishengir-compile`, or runtime scheduling.

## Expected behavior

A `tl.store(ptr, value)` followed by `tl.load(ptr)` in the same Triton program
must observe the just-written value in program order.

For `renormalize=True`, output top-k weights should be finite, sum to one,
match the PyTorch reference, and be stable across repeated launches.

### FL-wrapper result, 20 repetitions

| Implementation | `renormalize=False` | `renormalize=True` |
|---|---|---|
| Original single kernel | `ref_err=7.45e-09`, `repeat_diff=0`, ids correct | `sum_err=1.61e10`, `ref_err=3.18e9`, `repeat_diff=3.05e9`, ids correct |
| Split normalization into a second kernel | `ref_err=7.45e-09`, `repeat_diff=0`, ids correct | `sum_err=2.38e-07`, `ref_err=4.47e-08`, `repeat_diff=0`, ids correct |

Direct FlagGems calls show the same original-vs-split behavior.

## Minimal trigger: one output reload is enough

The diagnostic always preserves the original first-pass raw top-k stores. During
normalization, it reloads exactly the first `N` values from `output_ptr`; other
values are recomputed from the same logits.

| Number of `output_ptr` reloads | `renormalize=True` result |
|---:|---|
| 0 | `sum_err=2.38e-07`, `ref_err=5.96e-08`, `repeat_diff=0` |
| 1 | `sum_err=3.18e9`, `ref_err=3.18e9`, `repeat_diff=3.05e9` |
| 2 | `sum_err=5.94e9`, `repeat_diff=2.65e9` |
| 4 | `sum_err=1.12e10`, `repeat_diff=3.18e9` |
| 8 | `sum_err=2.05e10`, `repeat_diff=2.53e9` |

Ids are correct in every group. Therefore a single reload is sufficient to
trigger the failure.

## Direct capture of the loaded value

We added a diagnostic buffer that records the value returned by:

```python
val = tl.load(output_ptr + idx, mask=valid_rows)
```
before division and before the final writeback.
Over 20 repetitions:
capture_ref_err=5.08693408e+08
capture_repeat_diff=5.08693408e+08
output_ref_err=2.47692160e+09
output_repeat_diff=2.47692160e+09
The captured pre-division reload is already incorrect and nondeterministic.
Therefore the failure happens at the global-output reload, before division or
final output writeback.
## Relevant kernel pattern
First top-k loop writes raw weights.
```python
tl.store(output_ptr + rows * k + ki, curr_max, mask=valid_rows)
```
Later, in the same program:
```python
for ki in range(k):
    idx = rows * k + ki
    val = tl.load(output_ptr + idx, mask=valid_rows)
    tl.store(output_ptr + idx, val / norm, mask=valid_rows)
```
Removing this reload while retaining the initial raw stores is correct.
Performing the reload in a separate subsequent kernel is also correct.
## Compiler artifacts
The failing TTIR explicitly contains the required dependency:
```mlir
tt.store %output_ptr, %raw_weight, ...
...
%value = tt.load %output_ptr, ...
tt.store %output_ptr, %value_div_norm, ...
```
The corresponding TTAdapter Linalg MLIR represents the reload as:
```mlir
global output subview
→ memref.copy to local allocation
→ normalize
→ materialize/write back to the same global subview
```
The no-reload variant does not contain the corresponding output-buffer copy.

## Attached files

The attached archive is:

```text
flaggems-ascend-topk-correctness-issue-bundle.tar.gz
```

After extraction, it contains:

```text
FlagGems-topk-access-bisect/
├── artifacts/
│   ├── BUG_REPORT.md
│   └── matrix-split/
│       └── worktree.patch
└── tests/
    └── tmp_probe_fused_topk_matrix.py

FlagGems-topk-original/
├── artifacts/
│   ├── direct/
│   │   └── topk_softmax.py
│   └── cann-pass-bisect/
│       ├── kernel.o
│       ├── libkernel.so
│       ├── pass-ir.log
│       └── stdout.log
└── tests/
    └── tmp_probe_direct_topk_matrix.py

FlagGems-topk-noreload/
└── artifacts/
    └── direct/
        └── worktree.patch

FlagGems-topk-readback-count/
└── artifacts/
    └── readback-count/
        ├── n0.log
        ├── n1.log
        ├── n2.log
        ├── n4.log
        └── n8.log

FlagGems-topk-readback-capture/
└── artifacts/
    └── readback-capture/
        └── results.log

triton_cache_topk_dump_original/
└── -n4NncdV0g8G1ZspRQR34ttFID9CjzVduXt4x6LVY_g/
    ├── topk_gating_softmax_kernel.ttir
    └── topk_gating_softmax_kernel.ttadapter

triton_cache_topk_dump_noreload/
└── 27XlcmpAcu3c32BRBWKepaV0w9DYuxVXf1EaXga-KyQ/
    ├── topk_gating_softmax_kernel.ttir
    └── topk_gating_softmax_kernel.ttadapter
```

Purpose of each file
File(s) | Purpose
-- | --
BUG_REPORT.md | Full investigation summary, environment, results, and maintainer request.
tmp_probe_fused_topk_matrix.py | Minimal reproducer through the vLLM-Plugin-FL fused_topk wrapper.
tmp_probe_direct_topk_matrix.py | Minimal direct FlagGems reproducer.
original/.../topk_softmax.py | Original failing kernel source.
noreload/.../worktree.patch | Diagnostic variant that retains raw stores but avoids output_ptr reload.
matrix-split/worktree.patch | Correct two-kernel normalization workaround.
n0.log to n8.log | Demonstrates that one output reload is sufficient to trigger corruption.
readback-capture/results.log | Directly shows that the value returned by tl.load(output_ptr) is already corrupted before division/writeback.
*.ttir | Failing and working Triton IR.
*.ttadapter | Corresponding Ascend TTAdapter Linalg MLIR.
cann-pass-bisect/ | Manual CANN 8.5 bishengir-compile replay output and compiler logs.

## Request

Could you please help investigate the Ascend lowering/execution of this
same-program dependency, especially:

```text
global output write
→ global output reload into local memory
→ computation
→ writeback to the same global output
```

In particular, please confirm whether there may be a missing alias/dependency,
buffer-lifetime issue, or pipeline scheduling issue in the Ascend backend,
triton-adapter-opt, bishengir-compile, or runtime.

The split-kernel implementation is a reliable FlagGems workaround, but the
original same-program Triton semantics should be correct.

### Environment details

| Component | Version |
|---|---|
| Device target | `Ascend910B3` |
| CANN | `8.5.0` |
| Python | `3.11.14` |
| PyTorch | `2.8.0+cpu` |
| torch_npu | `2.8.0.post2` |
| Triton-Ascend | `3.2.0` |
| FlagGems baseline | `30bfcc073` / v5.0.2 |
| Reproducer shape | 64 tokens × 256 experts, bf16 logits, topk=8 |

[flaggems-ascend-topk-correctness-issue-bundle.tar.gz](https://github.com/user-attachments/files/30781754/flaggems-ascend-topk-correctness-issue-bundle.tar.gz)

**Comment by zhudada0120** [comment]
Hi @luentong , Thanks for the feedback. The issue has been successfully reproduced. We will post updates in this Issue as soon as we make progress.

**Comment by zhudada0120** [comment]
Hi @luentong , This issue has been resolved in the latest main‑branch build of the triton‑ascend package. We recommend switching to the newly‑compiled triton‑ascend package for verification.

**Comment by luentong** [comment]
Thanks for the update.

**Comment by hongziqi** [comment]
We will close this issue if there are no further concerns. Feel free to open a new issue if you encounter new problems later.
